core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -1254,6 +1255,13 @@ void cleanupExpiredRemoteLogSegments() throws RemoteStorageException, ExecutionE
                         canProcess = false;
                         continue;
                     }
+
+                    if (RemoteLogSegmentState.COPY_SEGMENT_STARTED.equals(metadata.state())) {
+                        // If the state is COPY_SEGMENT_STARTED and it's not under copying process, this must be the previously
+                        // failed copied state. We should clean it up directly.
+                        danglingSegments.add(metadata);
+                        continue;
```

Comment:
I think here we have a risk of race condition if `segmentIdsBeingCopied` is updated during the for loop which could cause the segment initially being copied to be deleted by mistake.
For the race condition to happen we would need to :
- When the list of segments is created L1244 a segment is currently being copied and present in `segmentIdsBeingCopied` 
- During the for loop, the segment copy finishes and its id is removed from `segmentIdsBeingCopied`
- L1259 `if (RemoteLogSegmentState.COPY_SEGMENT_STARTED.equals(metadata.state()))` can now be reached with state `COPY_SEGMENT_STARTED` for that segment because it has been removed from `segmentIdsBeingCopied` which would make it dangling and cause its deletion.

Reply:
That's a really good point! Let me fix it.

Reply:
PR updated in this commit: https://github.com/apache/kafka/pull/16959/commits/fc966752f36bb6769152ed92d85b8d97ada00003 . Also added test.